tracing-log/src/lib.rs

```diff
@@ -246,3 +256,136 @@ impl AsTrace for log::Level {
+    fn normalized_metadata(&'a self) -> Option<Metadata<'a>> {
+        let original = self.metadata();
+        if self.is_log() {
+            let mut fields = LogVisitor::new_for(self, level_to_cs(original.level()).1);
+            self.record(&mut fields);
+
+            Some(Metadata::new(
+                "log event",
+                fields.target.unwrap_or("log"),
+                original.level().clone(),
+                fields.file,
+                fields.line.map(|l| l as u32),
+                fields.module_path,
+                field::FieldSet::new(&["message"], original.callsite()),
+                Kind::EVENT,
+            ))
+        } else {
+            None
+        }
+    }
+
+    fn is_log(&self) -> bool {
+        // We can't use identify_callsite directly on Callsite as lazy_static
+        // breaks address comparison
+        self.metadata().callsite() == level_to_cs(self.metadata().level()).0.metadata().callsite()
+    }
+}
+
+struct LogVisitor<'a> {
+    target: Option<&'a str>,
+    module_path: Option<&'a str>,
+    file: Option<&'a str>,
+    line: Option<u64>,
+    fields: &'static Fields,
+}
+
+impl<'a> LogVisitor<'a> {
+    // We don't actually _use_ the provided event argument; it is simply to
+    // ensure that the `LogVisitor` does not outlive the event whose fields it
+    // is visiting, so that the reference casts in `record_str` are safe.
+    fn new_for(_event: &'a Event<'a>, fields: &'static Fields) -> Self {
+        Self {
+            target: None,
+            module_path: None,
+            file: None,
+            line: None,
+            fields,
+        }
+    }
+}
+
+impl<'a> Visit for LogVisitor<'a> {
+    fn record_debug(&mut self, _field: &Field, _value: &fmt::Debug) {}
+
+    fn record_u64(&mut self, field: &Field, value: u64) {
+        if field == &self.fields.line {
+            self.line = Some(value);
+        }
+    }
+
+    fn record_str(&mut self, field: &Field, value: &str) {
+        unsafe {
+            // The `Visit` API erases the string slice's lifetime. However, we
+            // know it is part of the `Event` struct with a lifetime of `'a`. If
+            // (and only if!) this `LogVisitor` was constructed with the same
+            // lifetime parameter `'a` as the event in question, it's safe to
+            // cast these string slices to the `'a` lifetime.
+            if field == &self.fields.file {
+                self.file = Some(&*(value as *const _));
+            } else if field == &self.fields.target {
+                self.target = Some(&*(value as *const _));
+            } else if field == &self.fields.module {
+                self.module_path = Some(&*(value as *const _));
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn log_call_site_is_correct() {
+        let record = log::Record::builder()
+            .args(format_args!("Error!"))
+            .level(log::Level::Error)
```

Comment:
it would be good to have a test like this for every level, just to make sure that they are all uniform.